File: 4CircularLL.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// Source code for linear linked list(insert,traverse,delete)


// defining the structure of Circular linked list
struct node{
    int data;
    struct node *next;
};
// initialising head,tail pointer to null
struct node *head=NULL;
struct node *tail=NULL;

// Function to create new node(used for insert function)
struct node* createNode(int data){  
   // assigning dynamic memory in heap for a new node and inserting data and next pointer as null.
     struct node *new_node = (struct node*)malloc(sizeof(struct node));
     new_node->data= data; 
     new_node->next = NULL;  
     return new_node; 
}

// Function to insert at the end of the list
void insertCircular(int data){

// creating a new node 
    struct node *new_node=createNode(data);

// checking if it is the first element to insert
    if (head==NULL){
        head=new_node;
        tail=new_node;
    }

// Adding new node after tail.
    else{
      tail->next=new_node;
      new_node->next=head; //new node pointing the head and hence folow the circular structure
      tail=tail->next;  // tail pointing to the newly inserted element.
    }
}
// Function to traverse the list and print from the begining.
void traverseList(){
     if(head==NULL){
        printf("Empty list"); //undeflow condition
        return;
    }
    struct node* tmp=head;
    // loop should stop whn we reach to the tail of ll
    while(tmp!=tail){ 
        printf("%d ",tmp->data);
        tmp=tmp->next;
    }
    // printing the last element
    printf("%d ",tmp->data);
    printf("%d ",tmp->next->data); // to check the circular property
    
    

}

// Recursive Function to delete the entire list
void deleteList(struct node* head)
{
  if (head == tail) {
  free(tail);
  return;}
  deleteList(head->next);
  free(head);

}
 // Main Program
 int main(){
     insertCircular(1);
     insertCircular(2);
     insertCircular(3);
     insertCircular(4);
     traverseList();
     deleteList(head);
 }

File: 4CircularLL.py
## template of circular ll
class createNode:
    def __init__(self,data):
        self.data=data
        self.next=None
## this class will contain all the methods and variables require for CLL        
class CircularLL:
    def __init__(self):
        ## initialising the head and tail pointers
        self.head=None 
        self.tail=None
 ## this method will insert at the end of cll   
    def insert(self,data):
        new_node=createNode(data)
        ## checking if first element to insert
        if self.head==None:
            self.head=new_node
            self.tail=new_node
        else:
            self.tail.next=new_node ## adding at the end of tail..This will  take O(1) time
            new_node.next=self.head ## pointing to the first node and hence maintaining the circular property
            self.tail=self.tail.next
        
## this method will traverse the cll and also verify the circular property
    def traverse(self):
        if(self.head==None):
            print("Empty List") ##underflow condition
            return
        tmp=self.head
        while(tmp!=self.tail): 
            print(tmp.data) 
            tmp=tmp.next
        print(tmp.data) ## printing the last data
        print(tmp.next.data) ## printing the first node through the tail node.
        

if __name__=="__main__":
    Cll=CircularLL()
    Cll.insert(1)
    Cll.insert(2)
    Cll.insert(3)    
    Cll.insert(4)
    Cll.traverse()
        
